### PR TITLE
Check for unterminated IF in yywrap

### PR DESCRIPTION
**Pull request by Rangi42**
This disallows starting/ending an IF inside an INCLUDEd file

### CODE REVIEW

**Comment by Rangi42** [comment]
Should this be a fatal error?

**Comment by ISSOtm** [comment]
I think that, instead of relying on a global variable, the context stack's `IF` depth should be the canonical one. This would:
- Avoid duplicating initialization
- Allow a greater amount of nesting (currently, nesting levels are global, whereas they'd be local to each context)
- Remove one global variable, which is always nice

I think it should be made into a fatal error.

**Comment by Rangi42** [comment]
So each context stack should store two numbers, the initial and current `IF` depth?

**Comment by ISSOtm** [comment]
Only its own `IF` depth. Each context would be created with a depth of 0, and only the current context's depth considered at any given time.

**Comment by Rangi42** [comment]
Done. This does mean that the lexer and parser have to use functions to read/update the context's `IF` depth.

**Comment by ISSOtm** [comment]
I don't think this should be a problem when using LTO. Should be enable it by default in the Makefile?

**Comment by Rangi42** [comment]
Yes, if `-O3` is default then LTO should be too.

**Comment by Rangi42** [comment]
(I'll leave that to you separately.)

**Comment by ISSOtm** [comment]
I'll add that separately; I just want to first check whether it's enabled in the CMakeLists as well.

**Comment by Rangi42** [comment]
Actually, it's just one `-flto` addition that applies to all programs, so it's here now. :P

File: src/asm/parser.y [hunk 1]
```diff
@@ -563,15 +563,15 @@ conditional	: if
...
+			fstk_IncIFDepth();
 			executeElseBlock = !$2;
 			if (executeElseBlock)
 				lexer_SetMode(LEXER_SKIP_TO_ELIF);
 		}
 ;
 
 elif		: T_POP_ELIF const T_NEWLINE {
-			if (nIFDepth <= 0)
+			if (fstk_GetIFDepth() <= 0)
```
Comment (ISSOtm) [inline src/asm/parser.y:574]: Since this returns a `uint32_t`, this should be `==`

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: test/asm/break.err [hunk 2]
```diff
@@ -2,4 +2,5 @@ warning: break.asm(9): [-Wuser]
     done 5
 warning: break.asm(17): [-Wuser]
     OK
-error: Unterminated IF construct (1 levels)!
+FATAL: break.asm(18) -> break.asm::REPT~1(23):
+    Ended block with 1 unterminated IF constructs
```
Comment (ISSOtm) [inline test/asm/break.err:6]: Do we want to bother with that `s`?

Comment (aaaaaa123456789) [reply test/asm/break.err:6]: Just do `construct(s)` if it's too hard to handle proeprly

Comment (Rangi42) [reply test/asm/break.err:6]: `constructs` is wrong for 1; `construct` is wrong for 2, 3, etc. `construct(s)`?

Comment (ISSOtm) [reply test/asm/break.err:6]:
```c
		fatalerror("Ended block with %" PRIu32 " unterminated IF construct%s\n",
			   contextStack->nIFDepth, contextStack->nIFDepth == 1 ? "" : "s");
```

**Review by ISSOtm — APPROVED** [review]

**Comment by ISSOtm** [comment]
We have a [`checkpatch` warning](https://github.com/gbdev/rgbds/pull/692/checks?check_run_id=1674363495#step:4:14), though.

**Comment by Rangi42** [comment]
Interesting; I've seen it complain before about `const char*` instead of `char const *`.

**Pull request merged**